internal/safebuffer/safereader/safereader.go

```diff
@@ -0,0 +1,160 @@
+package safereader
+
+import (
+	"bytes"
+	"io"
+	"log/slog"
+	"sync"
+)
+
+// Reader is a goroutine safe io.ReadCloser that makes initial data available to
+// callers as well as new data that arrives later via channel.
+type Reader struct {
+	mu        sync.Mutex
+	buf       *bytes.Buffer
+	close     func()
+	closed    <-chan struct{}
+	dataCh    <-chan []byte
+	dataAvail chan struct{}
+	closeOnce sync.Once
+}
+
+var _ io.ReadCloser = (*Reader)(nil)
+
+// jobIsRunning doesn't have a race because Buffer.NewReader calls it while
+// holding a lock that prevents new data, or the close of jobDone, from being
+// processed.
+func jobIsRunning(jobDone <-chan struct{}) bool {
+	select {
+	case <-jobDone:
+		return false
+	default:
+		return true
+	}
+}
+
+type Channels struct {
+	Data   chan []byte
+	Closed <-chan struct{}
+}
+
+// New returns a new Reader that keeps a copy of data and will append any data
+// that arrives on the returned Data channel. If the job is no longer running
+// when created, it is treated as a closed reader and no new data can be added.
+// Calls to Close() in this case are ignored.
+func New(data []byte, jobDone <-chan struct{}) (*Reader, *Channels) {
+	closed := make(chan struct{})
+	closeFn := func() { close(closed) }
+
+	buf := make([]byte, len(data))
+	copy(buf, data)
+
+	r := Reader{
+		close:  closeFn,
```

Comment:
Why is this a field on Reader?

Reply:
The func is "how" to close and signal to release resources, the chan is how the reader can know that it has been signaled to close to actually break out of blocking conditions and exit goroutines.